Adds feature to configure keycloak server with Scopes

mdanish98 opened the pull request:
This library is indeed really useful.
I was using this library to mock keycloak authorization token behavior. But one use case it was missing like when there is an implementation for getting authorization token with normal OAuth2 request using custom getAccessToken method. The scenario is like I have a project which has a method to provide authorization and it just takes clientId, clientSecret and tokenUrl and it has its own implementation to request JWT token. Here I have to check whether it is able to successfully get the access token and then able to provide access to the secured registry, and the secured registry checks for the required scopes needed to perform certain action (Read/Write etc).
So for this there is a need to configure the realm with clientScopes while configuring the server and by not using TokenConfig. 
I have added this feature to initialize and start KeyCloakMock server with clientScopes. I have also added the required test cases to cover this implementation.

Please let me know incase you need further information.

Thanks
Signed-off-by: Mohammad Ghazanfar Ali Danish <mdanish98@gmail.com>

mdanish98 commented:
Dear @ostrya

Did you had some time to check this PR ?
Actually I have another PR on my other project which is dependent on this. If this change is merged that would be easy to update the dependent PR.

Please let me know incase of any further information.

Thanks and Regards

ostrya reviewed (changes requested):
Sorry for the late reply. I like your change, but I have a suggestion so we can avoid manipulating static variables.

mdanish98 commented on mock/src/main/java/com/tngtech/keycloakmock/api/ServerConfig.java:
```diff
@@ -156,6 +157,31 @@ public Builder withPort(final int port) {
+     */
+    @Nonnull
+    public Builder withClientScopes(Set<String> scope) {
+      TokenConfig.addClientScopes(scope);
```
Thanks for your suggestions @ostrya 

As per your suggestions I have updated the pull request. Following are the changes I did:
- Added defaultScopes configuration in ServerConfig
- Removed withClientScopes method in ServerConfig
- Moved the openid scope from the TokenConfig to ServerConfig
- Removed the default value scope check from TokenConfigTest (default_values_are_used) as now default value of scope is configured via ServerConfig
- Modified the TokenGeneratorTest (config_is_correctly_applied) in such a way that it only checks the scope configured and not the defaultScope
- Now the defaultScope (openid) is added when Keycloakmock Server is instantiated
- Added additional test case in KeycloakMockTest to check if the server is started without using withDefaultScopes then by default it should contain 'openid' scope

As you know moving the default scope "openid" from TokenConfig to ServerConfig is a breaking change, so thatswhy I have to adapt few test cases inorder to fit with current changes.

Please let me know if you need further changes or if I misunderstood something.

mdanish98 commented on example-backend/bin/.gitignore:
```diff
@@ -0,0 +1,2 @@
+/main/
```
Thanks for your suggestions @ostrya

I have corrected this and updated the pull request.
Please let me know if you need further changes.

mdanish98 commented:
Dear @ostrya @mrmarbury @Achimh3011 @aaschmid ,

Actually we are dependent on this change and our project which is [Eclipse Basyx](https://github.com/eclipse-basyx/basyx-java-sdk) an **Industry 4.0** project is pending due to this pull request. 

I request you to please look into that change and merge it so that we can use it in our project.

ostrya commented:
Hi, sorry for the long delay. I had a really busy November and must then have lost track of this PR. I will have a look now.

mdanish98 commented on mock/src/main/java/com/tngtech/keycloakmock/api/ServerConfig.java:
```diff
@@ -118,15 +123,28 @@ public String getDefaultRealm() {
   }
 
+  /**
+   * The default scopes used in issuer claim.
```
Updated this in new commit

mdanish98 commented on mock/src/main/java/com/tngtech/keycloakmock/api/ServerConfig.java:
```diff
@@ -262,6 +280,20 @@ public Builder withResourceToMapRolesTo(@Nonnull String resource) {
+    /**
+     * Set default client scopes.
+     *
+     * <p>Set of client scopes to be configured. Default scope is 'openid'.
```
Updated this in new commit

mdanish98 commented on mock/src/main/java/com/tngtech/keycloakmock/api/TokenConfig.java:
```diff
@@ -221,9 +221,7 @@ public static final class Builder {
     @Nullable private String authenticationContextClassReference;
 
-    private Builder() {
-      scope.add("openid");
```
Updated the Javadoc in new commit

mdanish98 commented:
Dear @ostrya 

Many thanks for your suggestion I updated all the changes you requested.
One remark regarding the "please use a solution similar to the "resourcesToMapRolesTo" parameter:" comment:
As per your remarks I created a named singleton inside ServerModule. I tried to inject the named parameter into TokenGenerator's constructor but it was not working because the TokenGenerator is being used in SignatureComponent and due to this it was throwing error as [[Dagger/MissingBinding]](error: [Dagger/MissingBinding] @javax.inject.Named). So I thought that would be a major change because I might have to include the ServerModule.class in SignatureComponent like @Component(modules = {KeyModule.class, ServerModule.class}) . And thatswhy I did the same thing you suggested in TokenHelper class instead, i.e. how "resourcesToMapRolesTo" is also implemented.

Please let me know if this is fine otherwise I will do it in TokenGenerator.

Please review it and let me know for further changes.

Thanks and Regards

ostrya commented on mock/src/test/java/com/tngtech/keycloakmock/api/KeycloakMockTest.java:
```diff
@@ -29,4 +43,94 @@ void generated_token_uses_correct_issuer() throws Exception {
+
+    Set<String> scope = extractScopeFromToken(token);
+
+    assertThat(scope.size()).isEqualTo(3);
```
Cosmetic change:
```suggestion
    assertThat(scope).hasSize(3);
```

ostrya commented on mock/src/test/java/com/tngtech/keycloakmock/api/TokenConfigTest.java:
```diff
@@ -47,7 +47,6 @@ void default_values_are_used() {
     assertThat(config.getPreferredUsername()).isNull();
     assertThat(config.getRealmAccess().getRoles()).isEmpty();
     assertThat(config.getResourceAccess()).isEmpty();
-    assertThat(config.getScope()).isEqualTo("openid");
```
To document the new behavior, maybe assert the scope is empty by default? I.e.

```
assertThat(config.getScope()).isEmpty();
```

ostrya reviewed (commented):
Yes, the approach to inject the scopes is much better. And it's fine to inject the scopes in the token helper instead.

I have just two minor remarks on the tests, otherwise it looks good now.

ostrya commented:
I'll just merge and adapt the tests on the main branch.
